feat: Posting 기능 추가

madst0614 opened the pull request:
### 작업 내용
- Posting API 추가

### 변경 사항(추가시엔 추가사항)
- Post Controller 추가
- Post Service 추가
- Post Repository 추가

### 특이 사항
- jwtTokenProvider.getAccountFromToken(String token) 구현 필요
- Redis 서버 내 HashTagName 정보 조회 메소드 구현 필요(논의사항)
- Service 테스트 코드 작성 필요
- Controller 테스트 필요
- Service like, share 기능 PostingType 별로 엔드포인트 호출 과정 필요

### 테스트
- [ ] 테스트 코드
- [ ] api 테스트

### 관련 이슈(옵셔널)
- 없음

madst0614 committed:
feat: 엔티티(Posting, PostingTag, Tag) 및 필드 enum(PostingType) 추가

madst0614 committed:
chore: Querydsl 관련 Config 추가

madst0614 committed:
chore: Search 관련 Enums 추가

madst0614 committed:
refactor: Entity Class Naming 및 DB 변경에 따라 이름 수정

madst0614 committed:
add: Request 및 Search 기능 관련 Dto 추가

madst0614 committed:
add: PostingRepository - SearchByCondition 추가

예외 고려 및 테스트 필요
기타 기능 및 기타 Repository 추가 필요

madst0614 committed:
comment: SearchByCondition 주석 수정

madst0614 committed:
comment: SearchByCondition order 부분 수정

madst0614 committed:
add:HashTagRepository

madst0614 committed:
add: search by condition

madst0614 committed:
chore: PostingType 순서 변경

madst0614 committed:
add: 테스트 관련 PostingHashTagRepository 추가

madst0614 committed:
add: pagination, posting errorcode

madst0614 committed:
add: findhashidbyname 추가

madst0614 committed:
chore: setter 삭제

madst0614 committed:
refactor: searchbycondition ->findPostingPageByCondition

madst0614 committed:
refactor: searchbycondition ->findPostingPageByCondition

madst0614 committed:
add: 테스트 관련 PostingHashTagRepository 추가

madst0614 committed:
chore: 기타 코드 수정

madst0614 committed:
add: PostingService and Impl

예외 고려 및 테스팅 필요

madst0614 committed:
chore: 가독성 향상

madst0614 committed:
chore: 기타 주석 및 코드 수정

madst0614 committed:
chore: 좋아요 공유 기능 return 값 수정

madst0614 committed:
refactor: 필드 수정 및 @Autowired->RequiredArgsConstructor

madst0614 committed:
add : PostingServiceImpl test

madst0614 committed:
delete: file

madst0614 committed:
add: PostingController

madst0614 committed:
chore: 빋드 설정

madst0614 committed:
Merge branch 'dev' of https://github.com/7th-wanted-pre-onboarding-teamN/sns-feed into feature/posting

# Conflicts:
#	build.gradle
#	src/main/java/wanted/n/exception/ErrorCode.java

madst0614 committed:
chore : pull 중 수정

madst0614 committed:
rename: 함수 이름 수정

madst0614 committed:
add: of 함수 추가
request->condition

madst0614 committed:
add: PostingController

madst0614 committed:
add: Posting Controller 인증

madst0614 committed:
refactor: Dto->DTO 수정

madst0614 committed:
add: Service @Transactional 추가

HyunsooZo commented on src/main/java/wanted/n/service/impl/PostingServiceImpl.java:
```diff
@@ -0,0 +1,103 @@
+    @Autowired
+    PostingRepository postingRepository;
+    @Autowired
+    HashTagRepository hashTagRepository;
```
참고만 부탁드립니다!
해당부분은 저희가 사용중인 `lombok` 라이브러리를 사용해 
자동주입을 할 수도 있을 것 같습니다!
```java
@Service
@RequiredArgsConstructor
public class PostingServiceImpl implements PostingService {
    private final PostingRepository postingRepository;
    private final HashTagRepository hashTagRepository;
```

madst0614 replied:
수정했습니다!

HyunsooZo commented on src/main/java/wanted/n/service/impl/PostingServiceImpl.java:
```diff
@@ -0,0 +1,103 @@
+    }
+    @Override
+    public Posting getPostingDetail(Long postingId) {
+        Posting posting = postingRepository.findById(postingId).orElseThrow(()->new CustomException(ErrorCode.POSTING_NOT_FOUND));
```
제가 알기로는 `Service`<->`Controller` 간 데이터는 
데이터 응집성과 결합도 제어 및 보안상의 이유로 `DTO`로 감싸는게 권장되는 것으로 알고있습니다. 
현재 `GetPostingDetail` 메서드 에서는 엔티티를 바로 `Controller` 로 리턴하는것 같습니다!

madst0614 replied:
의견반영해서 DTO 추가했습니다!

HyunsooZo commented on src/main/java/wanted/n/service/impl/PostingServiceImpl.java:
```diff
@@ -0,0 +1,103 @@
+                .shareCount(posting.getShareCount()+1)
+                .build());
+        return null;
+    }
```
혹시 이렇게 구현하신 이유가 있는지 궁금합니다! 

1. 리턴타입이 `Boolean`인 이유가 있을까요? 
2. 기존 객체 필드값 업데이트가 아닌 다른 객체를 새로 생성해서 저장하시는 이유가 있을까요?

    2.1.  해당 케이스의 경우  데이터베이스에는 동일한 내용을 가진 두 개의 `Posting` 레코드가 생성될 것으로 보여요. 
아래와 같이 더티체킹을 사용해 저장하면 하나의 레코드만 관리하게 되어서 더욱 효율적이고 불필요한 데이터를 
추가적으로 저장할 필요가 없을 것 같아요! 

    2.2. 엔티티에 `@DynamicUpdate` 적용 시 엔티티의 변경 사항이 발생할 때 해당 변경 사항만을 업데이트하는 데 사용됩니다)
```java
@Override
@Transactional
 public void SharePosting(Long postingId) {
        Posting posting = postingRepository.findById(postingId)
                .orElseThrow(()->new CustomException(ErrorCode.POSTING_NOT_FOUND));
        posting.setShareCount(posting.getShareCount() + 1);
}
```

madst0614 replied:
1. void로 수정했었는데 깃헙이 미숙해서 반영이 안된거 같습니다!
2.1 제가 알기론 jpa save 기능이 id를 이용해서 우선 해당 데이터 객체를 찾고 없으면 추가 아니면 업데이트로 알고 있습니다! 따로 업데이트 메소드가 없는 것 같아서 save를 썼습니다!
2.2 제가 Entity는 Value Object로 해당 객체를 Read-only하게만 써서 객체를 다시 만들었습니다! Value Object는 내용이 달라지면(share_count+=1) 다른 객체로 취급해서 다시 만들어서 Save하는 과정을 사용했습니다! 

별개로, DB에 접근하는 객체는 Read only하게 사용하는게 좋다고 생각해서 Setter도 사용하지 않았습니다! 오직 데이터 변경은 서비스 단에서만 이뤄집니다!

HyunsooZo commented on src/main/java/wanted/n/service/impl/PostingServiceImpl.java:
```diff
@@ -0,0 +1,103 @@
+                .build());
+    }
+    @Override
+    public Boolean likePosting(Long postingId) {
```
이 메서드도 리턴타입이 `void` 가 되야하지 않을까 생각되는데 어떻게 생각하시나요?

madst0614 replied:
수정했습니다!

HyunsooZo commented on src/main/java/wanted/n/controller/PostingController.java:
```diff
@@ -1,13 +1,68 @@
+import javax.validation.Valid;
 
 @RequiredArgsConstructor
 @RequestMapping("/api/posting")
```
RestFul한 API 설계는 기본 복수형 사용을 권장하는 것으로 알고있는데 
어떻게 생각하시는지 궁금합니다! 
예 : 
포스팅 상세조회 `/api/postings/{postingId}`
포스팅 전체조회 `/api/postings`

madst0614 replied:
제가 RestFul 부분을 놓쳐서 수정했습니다!

HyunsooZo commented on src/main/java/wanted/n/controller/PostingController.java:
```diff
@@ -1,13 +1,68 @@
+    public ResponseEntity<Void>  likePosting(@PathVariable("id") Long id) {
+        postingService.likePosting(id);
+
+        return new ResponseEntity<>(HttpStatus.OK);
```
> 멘토_안병훈 — 2023.10.25. 오전 10:55
> 유저 >. (PUT, PATCH) > 서버 > (POST) > SNS 서버
> 유저가 저희 서버에 요청할때에는 PUT 또는 PATCH 로 쏘겠지만,
> 저희 서버 > SNS 서버 에 요청할때에는 좋아요 record가 생성 될거라 POST 로 말씀드렸습니다.
> 좋아요, 공유하기에 있는 endpoint 는 저희 API 가 아닌 외부 SNS API 스펙(가정) 입니다~! 

위 요건을 보면 해당 메서드는 `POST`메서드를 사용하는게 맞지 않을까 싶습니다! 
그리고 리소스 일부 수정은 `PUT` 이 아닌 `PATCH`를 사용하는게 좀더 restful하다고 알고있습니다! 
참고부탁드려요~

madst0614 replied:
해당 호출이 성공할 시 response status 200 해당 게시물의 like_count가 1 증가합니다.
외부 요청과 달리 저희 쪽 요청은 Posting Column의 like_count, share_count가 업데이트 되기에 Http 업데이트 메소드가 맞다고 생각합니다! PUT보다는 PATCH가 더 맞는 것 같아 수정했습니다

추가 설명이 좀 부족한 감이 있는데, 저희 쪽 로직이
요청 -> 컨트롤러 -> 서비스 -> 외부 서비스 좋아요 요청 후(해당 서버 api 인터페이스에 따름), 우리 데이터베이스 업데이트 입니다!

HyunsooZo commented on src/main/java/wanted/n/dto/PostingSearchConditionDto.java:
```diff
@@ -0,0 +1,24 @@
+@NoArgsConstructor(access = AccessLevel.PROTECTED)
+@AllArgsConstructor
+@Builder
+public class PostingSearchConditionDto {
```
Json직렬화가 필요없는 DTO의 경우 `@Builder`를 사용하신다면 따로 생성자를 호출하는 코드가 없는 한 
`@NoArgsConstructor`,`@AllArgsConstructor` 어노테이션은 굳이 붙일 필요는 없다고 알고 있습니다. 
이부분도 참고만 부탁드려요~

HyunsooZo reviewed (dismissed):
전체적으로 깔끔하게 잘 작성된 코드로 보입니다! 
고생 많으셨어요! 🫡

다만 커밋을 클래스 단위가 아닌 기능단위로 해주시면 해당 기능에 
필요한 로직 및 코드만을 한 커밋 내에서 리뷰할 수 있고 이슈트래킹에도 더 유리하여
코드 퀄리티를 높이는데에도 더 좋을 것 같습니다! 
(예: 포스팅 검색기능 추가 , 포스팅 상세정보확인 기능 추가 ....)

그리고 완성되지 않은 기능의 경우는 커밋하지 않는것이 좋다고 알고있어요!
깃은 코드베이스 상태를 기록하고 변경 내용을 추적하는 데 사용되는 도구이므로, 
미완성 또는 작동하지 않는 코드를 커밋하면 다른 팀원에게 혼란을 초래하거나 가독성 저해
이슈추적 어려움 등의 문제가 생길 수 있다고 알고있습니다. 참고부탁드립니다! 

그리고 저희 코드 컨벤션에 맞춘 커밋 타이틀 사용해주시면 더 좋을 것 같아요! 
예:  
기능추가,수정 ->  `feat: ~~~기능 추가 또는 수정`
파일 삭제 -> `remove: ~~~파일 삭제`
코드 포맷 -> `style: ~~가독성을 위한 줄바꿈`
빌드업무(예: 의존성추가등..) -> `chore: ~~의존성 추가`
테스트코드 추가 수정 -> `test: ~~~테스트 코드 추가 또는 수정 `

Team"N" [코드 컨벤선 링크](https://www.notion.so/6950ab1c30c44ffa8c3d2c20ccff872b?pvs=4)

참고부탁드려요 다시한번 고생많으셨습니다!! 🫡

madst0614 committed:
refactor: request / response 관련 dto 추가

madst0614 committed:
test: PostingService 테스트 추가

madst0614 committed:
refactor: posting cnt 컬럼들 nullable 추가 및 기타 리팩토링

madst0614 committed:
refactor: getPostingDetail -> upPostingViewCount

upPostingViewCount만으로 getPostingDetail 기능을 수행할 수 있고
id를 이용해서 posting만 가져오는 경우는 JPA의 findbyId로 대체 가능

madst0614 committed:
bug: 의존성 주입 문제 및 뷰카운트 증가 안하던 버그 수정

junkyu92 commented on src/main/java/wanted/n/repository/custom/impl/HashTagRepositoryCustomImpl.java:
```diff
@@ -0,0 +1,21 @@
+
+    @Override
+    public Optional<HashTag> findByName(String name) {
+        return Optional.ofNullable(queryFactory.selectFrom(hashTag).where(hashTag.name.eq(name)).fetchOne());
```
참고만 해주세요.
Spring Data Jpa에서는 이런 간단한 기능은 커스텀해서 구현할 필요없이 네이밍으로만 기능을 만들 수 있습니다. 
현수님이 작성하신 UserRepository참고하시면 될 것 같습니다.

madst0614 replied:
감사합니다! 알려주신 대로 구현 적용해서 해보겠습니다!

junkyu92 reviewed (approved):
전체적으로 깔끔하게 작성하신 것 같습니다.

12hyeon reviewed (approved):
위 논의사항을 잘 반영해주신 것 같습니다!
